src/probnum/linops/_linear_operator.py

```diff
@@ -777,6 +778,22 @@ def trace(self) -> np.number:
 
         return self._trace_cache
 
+    def _diagonal(self) -> np.ndarray:
+        """Diagonal of the linear operator.
+
+        You may implement this method in a subclass.
+        """
+        return np.diagonal(self.todense(cache=False))
```

Comment:
How about we modify the default implementation of `trace` (a loop that iteratively multiplies with canonical unit vectors) to compute the diagonal instead and then we can implement `trace` as `np.sum(self.diagonal)` by default? This would mean O(n) memory complexity in both cases and we can leverage the diagonal cache in both cases.

Reply:
It's O(n^2) memory complexity then, right? Because the default implementation of `diagonal` computes the dense representation of the linop. This might be a problem for large matrix-free linops where no memory-efficient implementation of `diagonal` is available. (To be fair, I can't think of an example right now - anytime I think of large linops I think of covariance matrices where computation of the diagonal is indeed possible with O(n) memory.)
But from an implementation POV, I agree that it would be more elegant.

Just wanted to raise that point, lmk what you think